### ISSUE TITLE
XRefs between chapters don't work the same in HTML, EPUB and PDF

### ISSUE THREAD

**Issue by agoncal** [asciidoctor/asciidoctor-epub3#136]
The code of this sample is available at : https://github.com/agoncal/agoncal-other/tree/master/04-AsciidocXrefStyle

Let's say I have a spine document with two chapters and one introduction. This introduction "talks" about the chapters, so, it needs to relate to those chapters (using Xref):

```
= XRef Between Chapters
:doctype: book
:sectnums:
:toc:

include::introduction.adoc[leveloffset=+1]

include::chapter-one.adoc[leveloffset=+1]

include::chapter-two.adoc[leveloffset=+1]
```

## Works on HTML and PDF (not on EPUB)

To make the XRefs work on HTML and PDF, I can use either way these two syntaxes. `[[chapter-one]]` for the chapter, and `<<chapter-one>>` for the reference:

```
[[chapter-one]]
= First chapter with listing
```

```
[[chapter-two]]
= Second chapter with table
```

```
= Introduction

* <<chapter-one>>
* <<chapter-two>>
```

This syntax also works 

```
[#chapter-one]
= First chapter with listing

[#chapter-two]
= Second chapter with table
```

## Works on EPUB as well

But this syntax does not work in EPUB. So has stated in [EPUB composition rules](https://github.com/asciidoctor/asciidoctor-epub3#composition-rules) I've used another syntax that works on all 3 formats:

```
[chapter-one#chapter-one]
= First chapter with listing
```

```
[chapter-two#chapter-two]
= Second chapter with table
```

```
= Introduction

* <<chapter-one#chapter-one>>
* <<chapter-two#chapter-two>>
```

But when I do that, I lose the numbering ('Chapter 1.', 'Chapter 2.'...)

I'm a bit confused with the different syntaxes (`[[chapter-one]]`, `[#chapter-one]`, `[chapter-one#chapter-one]`) and the way numbering works.

## Related issues

* https://github.com/asciidoctor/asciidoctor-epub3/issues/27
* https://github.com/asciidoctor/asciidoctor-epub3/pull/94
* https://github.com/asciidoctor/asciidoctor/issues/2397
* https://github.com/asciidoctor/asciidoctor-epub3/issues/132

**Comment by mojavelinux** [comment]
Duplicate of #132, which is now resolved. I'll double check, though, to make sure I haven't missed anything. Then I'll make a release.

**Comment by mojavelinux** [comment]
Here's the correct way to structure the document.

**book.adoc**

```
= XRef Between Chapters
:doctype: book
:sectnums:
:toc:
:xrefstyle: full

include::introduction.adoc[leveloffset=+1]

include::chapter-one.adoc[leveloffset=+1]

include::chapter-two.adoc[leveloffset=+1]
```

**introduction.adoc**

```
[preface]
= Introduction

* <<chapter-one.adoc#chapter-one>>
* <<chapter-two.adoc#chapter-two>>
```

**chapter-one.adoc**

```
[#chapter-one]
= First chapter with listing
```

**chapter-two.adoc**

```
[#chapter-two]
= Second chapter with table
```

The form `[#chapter-two]` assigns the ID "chapter-two" to the section.

The form `<<chapter-two.adoc#chapter-two>>` creates a cross reference to the ID "chapter-two" in the source file "chapter-two.adoc". The reason `#chapter-two` is required here is because in a PDF, you need to have a concrete anchor point or else it will link to the top of the document. (At least, that's a workaround that's necessary right now).

Unfortunately, an xref to a chapter in EPUB3 won't start with "Chapter N". That's because in EPUB3, chapter numbers are not used atm.

**Comment by mojavelinux** [comment]
So you are correct, xrefs between chapters in EPUB3 don't work the same as in other formats. That's because in EPUB3, we're actually creating separate documents and core doesn't yet support the xrefstyle for such references.

**Comment by agoncal** [comment]
@mojavelinux I managed to make all the references work, it's the xrefstyle that doesn't work (e.g. `:xrefstyle: short`, which is what I use).

> core doesn't yet support the xrefstyle for such references.

Dan, so you are saying that `:xrefstyle: short` is not implemented in EPUB, right ? Any idea of when this will be implemented?

**Comment by mojavelinux** [comment]
I'm saying that when you reference one chapter from another chapter, it will just show the regular chapter title (not a fancy one). The only time you get the xrefstyle is when you refer to a block with a title inside the same chapter. For example, if you refer to an image with a title, it will say "Figure N".

I'm actually not sure how to solve the chapter-to-chapter problem without going back to core and adding styled xreftext to the document itself. But then we also have to figure out some way to know it's a chapter. I have to do some thinking about it.

**Comment by mojavelinux** [comment]
To clarify, the xrefs won't be broken. They just won't be fancy. That's because we don't really have such a notion in EPUB3. But they will work correctly for PDF and HTML5.

**Comment by agoncal** [comment]
Here we are focusing on chapters, but the same problem happens with Listings, Figures and Tables. If I have the following:

```
:xrefstyle: short

My reference to <<validating-constraints>> does not display Listing 1 un EPUB

[[validating-constraints]]
.Order Cascading Validation to Its Order Lines
[source]
----
blah blah
----
```

In HTML and PDF this translates to "My reference to Listing 1" but in EPUB it's "My reference to Order Cascading Validation to Its Order Lines". So `:xrefstyle: short` won't work for Listings in EPUB.

**Comment by mojavelinux** [comment]
Once Asciidoctor EPUB3 is released, those should work. Are you testing the current release or master?

**Comment by mojavelinux** [comment]
I just checked against master and it works.

The release was being held up by #157. I just submitted a fix for that, so I should be able to release shortly.

**Comment by agoncal** [comment]
@mojavelinux really !?! It would be awesome if you could release it, that would be great !

**Comment by mojavelinux** [comment]
I'll be sure to ping you when it's out!

**Comment by mojavelinux** [comment]
As this issue is focused on chapters, it will carry over to alpha.9. The alpha.8 release should address all the in-chapter references.

**Comment by agoncal** [comment]
@mojavelinux That means that Listing, Table and Figure will work on :xrefstyle: short for Alpha8, but the chapters only in Alpha9, right ?

**Comment by mojavelinux** [comment]
Correct.

**Comment by mojavelinux** [comment]
I forgot, there is a workaround for this. You can set the `docreftext` attribute in the chapter document:



**chapter-1.adoc**
```
= Name of Chapter
:docreftext: Chapter 1
```

That value will be used in place of the doctitle when referencing the chapter.

**another-chapter.adoc**
```
See <<chapter-1.adoc#>>
```

You can even use a preprocessor conditional to conditionally set the docreftext attribute (though, in your case, you probably don't need this formality).

```
= Name of Chapter
ifeval::["{xrefstyle}"=="short"]
:docreftext: Chapter 1
endif::[]
```

The only downside is that you have to number the chapters manually.

**Comment by mojavelinux** [comment]
What we're looking for to complete this feature is to have this capability without having to set the docreftext attribute explicitly.

**Comment by agoncal** [comment]
@mojavelinux That's a great workaround. I'm using it for all my chapters and that does the trick. Thanks Dan

**Comment by mojavelinux** [comment]
That's great to hear! Thanks for following up, @agoncal!

**Comment by pilhuhn** [comment]
It would be good if `data/samples/sample-book.adoc` would show the options.
Also I am currently trying to allow MicroProfile-Metrics spec to be rendered as epub, which requires a major restructuring due to the limitations here. 

Shouldn't anchors be unique across a document? In that case asciidoctor could read the file, create the ast wiht the anchors in memory and then resolve those without the need to reference chapters and/or input files.
Also for explicitly creating an anchor with the name of the input file: as the input is read via `include::[]` directive, the name should be known and such an anchor could be implicitly created (?)

**Comment by slonopotamus** [comment]
I don't think there will be any progress on this issue until we implement #47.
